packages/core-contracts/contracts/proxy/UUPSImplementation.sol

```diff
@@ -6,9 +6,31 @@ import "../common/CommonErrors.sol";
 import "./ProxyStorage.sol";
 
 contract UUPSImplementation is ProxyStorage, ContractUtil, CommonErrors {
+    error SterileImplementation(address implementation);
+    error SimulatedUpgradeFailed();
+    error UpgradeToNotCalledViaProxy();
+
     event Upgraded(address implementation);
 
-    function _upgradeTo(address newImplementation) internal virtual {
+    address private immutable __self = address(this);
+
+    // WARNING!!!
```

Comment:
I wasn't able to not expose the critical upgradeTo functions here because the anti-bricking mechanism needs the interface to exist.

I know that our rule is to never expose critical functions by default, but so far I can't think of a way to commit to the rule here.